zenoh/src/api/builders/close.rs

```diff
@@ -0,0 +1,182 @@
+//
+// Copyright (c) 2024 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+use std::{
+    future::{Future, IntoFuture},
+    pin::Pin,
+    time::{Duration, Instant},
+};
+
+use async_trait::async_trait;
+use tokio::{select, time::sleep};
+use tracing::warn;
+use zenoh_core::{Resolvable, Wait};
+use zenoh_result::ZResult;
+use zenoh_runtime::ZRuntime;
+use zenoh_task::TaskController;
+
+/// A behavior selector for close operations used if initial timeout was reached
+pub enum CloseBackoff {
```

Comment:
I think the term `backoff` is misused here. It seems to be a normal timeout.

Reply:
Well, this is a policy describing what to do if the initial timeout was not enough, that's why I used `backoff` term. The idea is that user might have some requirements to have typical finalization delay that will become an initial timeout and some absolute maximum delay that can be described through `CloseBackoff::Wait(_)`

Reply:
If the timeout was not enough, then why bother? If the timeout was too short, then it needs to be increased. I don't see really the use case nor the need for such added complexity in the API.

Reply:
And at the moment of switching in between the initial and backoff timeouts a warning message will be printed out to the log. Currently logging here is very minimal, but in the future I want to extend it.

Reply:
Again, why the use should set 2 timeouts? The use case and the need is unclear.

Reply:
To give user warning in case if the typical timeout failed - that can be used to early catch finalization delay problems etc - but on the other hand to give the ability to have extended timeout for proper finalization completion to avoid having semi-finalized `Session`.

Reply:
I don't think it's a real need nor we have a real use case that requires that. 
Can you provide one? Otherwise I believe a simple timeout will suffice.

Reply:
Started describing the use-case, but then I understood that everything of this can be implemented in the user code. I will fall back to the simple timeout then :)